From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp60: Implement the argument-less facet _Getcat functions.

Fast mode for msvcp60: VC6's argument-less static _Getcat of the facets
(locale::facet, collate, ctype, codecvt, numpunct, num_get, num_put,
time_get, time_put) were spec stubs; they return the facet's category
(-1 for the facet base class).
---
 dlls/msvcp60/msvcp60.spec | 68 +++++++++++++++++++++++------------------------
 dlls/msvcp90/locale.c     | 16 +++++++++++
 2 files changed, 50 insertions(+), 34 deletions(-)

diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index 830c32e..9e12872 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -2656,18 +2656,18 @@
 @ cdecl -arch=win64 ?_Getcat@?$_Mpunct@D@std@@SA_KXZ() money__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$_Mpunct@G@std@@SAIXZ() money__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$_Mpunct@G@std@@SA_KXZ() money__Getcat_old
-@ stub -arch=win32 ?_Getcat@?$codecvt@DDH@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$codecvt@DDH@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$codecvt@GDH@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$codecvt@GDH@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$collate@D@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$collate@D@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$collate@G@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$collate@G@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$ctype@D@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$ctype@D@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$ctype@G@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$ctype@G@std@@SA_KXZ
+@ cdecl -arch=win32 ?_Getcat@?$codecvt@DDH@std@@SAIXZ() ctype__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$codecvt@DDH@std@@SA_KXZ() ctype__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$codecvt@GDH@std@@SAIXZ() ctype__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$codecvt@GDH@std@@SA_KXZ() ctype__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$collate@D@std@@SAIXZ() collate__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$collate@D@std@@SA_KXZ() collate__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$collate@G@std@@SAIXZ() collate__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$collate@G@std@@SA_KXZ() collate__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$ctype@D@std@@SAIXZ() ctype__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$ctype@D@std@@SA_KXZ() ctype__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$ctype@G@std@@SAIXZ() ctype__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$ctype@G@std@@SA_KXZ() ctype__Getcat_noarg
 @ cdecl -arch=win32 ?_Getcat@?$messages@D@std@@SAIXZ() messages__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$messages@D@std@@SA_KXZ() messages__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$messages@G@std@@SAIXZ() messages__Getcat_old
@@ -2688,28 +2688,28 @@
 @ cdecl -arch=win64 ?_Getcat@?$moneypunct@G$00@std@@SA_KXZ() money__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$moneypunct@G$0A@@std@@SAIXZ() money__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$moneypunct@G$0A@@std@@SA_KXZ() money__Getcat_old
-@ stub -arch=win32 ?_Getcat@?$num_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$num_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$num_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$num_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$numpunct@D@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$numpunct@D@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$numpunct@G@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$numpunct@G@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@?$time_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@?$time_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KXZ
-@ stub -arch=win32 ?_Getcat@facet@locale@std@@SAIXZ
-@ stub -arch=win64 ?_Getcat@facet@locale@std@@SA_KXZ
+@ cdecl -arch=win32 ?_Getcat@?$num_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIXZ() numeric__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$num_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KXZ() numeric__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$num_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIXZ() numeric__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$num_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KXZ() numeric__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIXZ() numeric__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KXZ() numeric__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIXZ() numeric__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KXZ() numeric__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$numpunct@D@std@@SAIXZ() numeric__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$numpunct@D@std@@SA_KXZ() numeric__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$numpunct@G@std@@SAIXZ() numeric__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$numpunct@G@std@@SA_KXZ() numeric__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIXZ() time__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KXZ() time__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIXZ() time__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KXZ() time__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIXZ() time__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KXZ() time__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@?$time_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIXZ() time__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@?$time_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KXZ() time__Getcat_noarg
+@ cdecl -arch=win32 ?_Getcat@facet@locale@std@@SAIXZ() locale_facet__Getcat_noarg
+@ cdecl -arch=win64 ?_Getcat@facet@locale@std@@SA_KXZ() locale_facet__Getcat_noarg
 @ cdecl -arch=arm ?_Getcoll@_Locinfo@std@@QBA?AU_Collvec@@XZ(ptr ptr) _Locinfo__Getcoll
 @ thiscall -arch=i386 ?_Getcoll@_Locinfo@std@@QBE?AU_Collvec@@XZ(ptr ptr) _Locinfo__Getcoll
 @ cdecl -arch=win64 ?_Getcoll@_Locinfo@std@@QEBA?AU_Collvec@@XZ(ptr ptr) _Locinfo__Getcoll
diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index 22c6cce..07af0b3 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -3560,6 +3560,22 @@ static const mpunct *money_use_mpunct(const locale *loc, BOOL intl, int type)
 }
 #endif
 
+#if _MSVCP_VER < 70
+/* VC6's argument-less _Getcat: the category of the facet, -1 for the
+ * facet base class */
+size_t __cdecl locale_facet__Getcat_noarg(void) { return -1; }
+size_t __cdecl collate__Getcat_noarg(void) { return LC_COLLATE; }
+size_t __cdecl ctype__Getcat_noarg(void) { return LC_CTYPE; }
+size_t __cdecl numeric__Getcat_noarg(void)
+{
+#ifdef SG_MUTANT_GETCAT60_NUMERIC_IS_TIME
+    return LC_TIME;
+#endif
+    return LC_NUMERIC;
+}
+size_t __cdecl time__Getcat_noarg(void) { return LC_TIME; }
+#endif
+
 /* ?id@?$collate@D@std@@2V0locale@2@A */
 locale_id collate_char_id = {0};
 
